# 01 — Data Understanding

Goal: understand the dataset before building any model — what data we have, what each column means, and what problems exist.

## 1. Import libraries

In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)

## 2. Load the dataset

In [2]:
df = pd.read_csv("../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv")
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


## 3. Dataset size

In [3]:
df.shape

(7043, 21)

*Fill in after running:*
- Number of customers (rows): 
- Number of features (columns): 

## 4. Column names

In [4]:
df.columns.tolist()

['customerID',
 'gender',
 'SeniorCitizen',
 'Partner',
 'Dependents',
 'tenure',
 'PhoneService',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaperlessBilling',
 'PaymentMethod',
 'MonthlyCharges',
 'TotalCharges',
 'Churn']

Group the columns mentally into categories, e.g.:
- **Identifier:** customerID
- **Demographics:** gender, SeniorCitizen, Partner, Dependents
- **Account info:** tenure, Contract, PaperlessBilling, PaymentMethod
- **Services:** PhoneService, MultipleLines, InternetService, OnlineSecurity, OnlineBackup, DeviceProtection, TechSupport, StreamingTV, StreamingMovies
- **Billing:** MonthlyCharges, TotalCharges
- **Target:** Churn

## 5. Data types

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


**Watch out:** `TotalCharges` will show up as `object` (text), not a number.
That's a red flag worth investigating below — it's not actually a text field.

## 6. Investigate the TotalCharges dtype issue

In [6]:
# TotalCharges looks numeric but is stored as text — find out why
non_numeric = df[pd.to_numeric(df['TotalCharges'], errors='coerce').isna()]
non_numeric[['customerID', 'tenure', 'MonthlyCharges', 'TotalCharges']]

,customerID,tenure,MonthlyCharges,TotalCharges
488,4472-LVYGI,0,52.55,
753,3115-CZMZD,0,20.25,
936,5709-LVOEQ,0,80.85,
1082,4367-NUYAO,0,25.75,
1340,1371-DWPAZ,0,56.05,
3331,7644-OMVMY,0,19.85,
3826,3213-VVOLG,0,25.35,
4380,2520-SGTTA,0,20.00,
5218,2923-ARZLG,0,19.70,
6670,4075-WKNIU,0,73.35,


These rows have a blank/whitespace string instead of a number — usually customers
with `tenure == 0` (brand new, so no total charges yet). This is why `df.isnull().sum()`
below will report 0 missing values for TotalCharges even though these blanks exist —
they're empty strings, not `NaN`. We will fix the dtype and decide how to handle these
rows during cleaning (Day 3), not today.

## 7. Missing values

In [7]:
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

**Note:** this will likely show 0 for every column, including `TotalCharges` —
see the investigation above for why that's misleading.

## 8. Duplicates

In [8]:
df.duplicated().sum()

np.int64(0)

In [9]:
# customerID should be unique per customer — check it explicitly too
df['customerID'].duplicated().sum()

np.int64(0)

## 9. Summary statistics

In [10]:
df.describe()

,SeniorCitizen,tenure,MonthlyCharges
count,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692
std,0.368612,24.559481,30.090047
min,0.000000,0.000000,18.250000
25%,0.000000,9.000000,35.500000
50%,0.000000,29.000000,70.350000
75%,0.000000,55.000000,89.850000
max,1.000000,72.000000,118.750000


In [11]:
df.describe(include='object')

,customerID,gender,Partner,Dependents,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,TotalCharges,Churn
count,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043
unique,7043,2,2,2,2,3,3,3,3,3,3,3,3,3,2,4,6531,2
top,3186-AJIEK,Male,No,No,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,20.2,No
freq,1,3555,3641,4933,6361,3390,3096,3498,3088,3095,3473,2810,2785,3875,4171,2365,11,5174


Look at:
- Average / min / max for `tenure` and `MonthlyCharges`
- Any values that look like outliers or data-entry errors
- For categorical columns: how many unique values, and the most frequent one

## 10. Target variable — Churn

In [12]:
df["Churn"].value_counts()

Churn
No     5174
Yes    1869
Name: count, dtype: int64

In [13]:
df["Churn"].value_counts(normalize=True) * 100

Churn
No     73.463013
Yes    26.536987
Name: proportion, dtype: float64

**Note the class imbalance** (this dataset is typically ~73% No / ~27% Yes).
This matters later: accuracy alone will be a misleading metric for the model,
and you may need techniques like class weighting, SMOTE, or choosing metrics
like recall/F1/ROC-AUC instead of plain accuracy.

## 11. Initial Observations

*(Replace this with your own actual findings after running everything above.)*

- The dataset contains customer demographic, account, service, and billing information.
- The target variable is `Churn` (Yes/No), and it is imbalanced (~X% Yes / ~Y% No).
- `TotalCharges` is stored as text due to blank values for customers with 0 tenure — needs conversion to numeric during cleaning.
- No true duplicate rows / duplicate customerIDs found (confirm with your actual output).
- Several categorical features will need encoding before modeling.
- `customerID` is a unique identifier and should be dropped before modeling.
